[design-docs] Add opmodes design doc

PeterJohnson opened the pull request:
This document describes a standardized approach for operators to select different code to run for different robot modes of operation (auto/teleop/test) and for programmers to easily write code that creates these selection options.

TODO:
- [x] Add migration sections for FTC and FRC summarizing how this is different than what is used in 2025-6
- [ ] Resolve open trades
- [ ] Add more detail to C++ and Python sections, particularly Python
- [x] Split command-based to separate design doc as its structure is substantially different
- [ ] Improve examples (especially command-based, as this will be probably the one most commonly used)
- [ ] Develop implementation to see how this really looks across examples and templates and work through any implementation issues that might affect the user facing API

PeterJohnson committed:
[design-docs] Add modes design doc

PeterJohnson committed:
Improve command autos example a bit

PeterJohnson committed:
Wrap loadComplexPath function call in a runOnce

PeterJohnson committed:
Add some trades

PeterJohnson committed:
Add trade re: command-based modes

PeterJohnson committed:
Tweak wording

jwbonner commented:
Here are some thoughts from my initial read of things.

1. How do default commands work (or how would they be replaced)? The command-based example shows binding a joystick control command with `whileTrue`, which would produce a different behavior (it is not automatically rescheduled when another command ends). Bringing back `whileTrueContinuous` isn't quite right either since it would require all other commands on the subsystem to be uninterruptible. Seems like default commands were a fitting solution, but I'm not sure how they would be translated to this structure.

2. My understanding is that the `CommandScheduler` periodic call would be triggered automatically when running a command-based mode (it would previously have been part of `robotPeriodic`). It seems like there are many uses cases where allowing the command scheduler to *not* run (under any circumstance) would be quite problematic: subsystem periodic methods are frequently used for telemetry, verifying the state of mechanisms for safety, or enforcing other overrides. For AdvantageKit as one such use case, it is critically important that subsystem `periodic` methods are *always* called for telemetry and replay. I think there is also a role for the global `robotPeriodic` for non-subsystem mode-agnostic telemetry (as well as other global tasks like updating LEDs), though I don't know how to make that fit with the requirement to support linear modes (or periodic modes of different frequencies). There's also a risk of users trying to run logic in `robotPeriodic` that does not belong there, which would need to be addressed.

I think there has been some discussion of this already, but a partial solution I could see to (1) and (2) is having a separate robot base class for command-based that *requires* only command-based modes and calls the command scheduler itself across all modes (and maybe provides a global `robotPeriodic` callback for other telemetry, possibly integrated more deeply with the command scheduler). Arguably there's still an application for `robotPeriodic` telemetry even without command-based, but maybe that's just impractical to support (and should have some other proposed alternative for global telemetry and periodic operations).

3. Is the plan that the `RobotBase` class managing these modes is single-threaded or multi-threaded? (Seems like maybe the latter if it needs to deal with running and killing linear modes). My concern here is designing things in a way that it can be made AdvantageKit-compatible in a fairly straightforward way (i.e. ideally without completely reimplementing `RobotBase`). For example, it would be easy for us to simply require AdvantageKit users to not use linear modes as long as `RobotBase` is otherwise deterministic (single-threaded?) when running periodic and command modes. This could potentially also be addressed with a separate command-based robot base (which would be closer to the current `TimedRobot`), but we don't necessarily want to tie all AdvantageKit users to command-based either.

Clearly I'm approaching this (in part) from the perspective of supporting AdvantageKit in some form going forward. While I don't think we're planning any _official_ AdvantageKit integration in 2027 (maybe 2028 or beyond), we have also been successful on the roboRIO with tweaking WPILIb to minimize the number of "intrusive" modifications required on the part of AdvantageKit. I'm operating on the assumption that we would like to continue a similar approach in 2027 to minimize the support burden for FRC teams that continue to use AdvantageKit.

KangarooKoala reviewed (commented):
These are mostly just comments on clarity and grammar. (Up to and not including the design section, since that's as far as I got before needing to go) There's a few more nitpicks (inconsistent usage of dictionary versus map, inconsistent inclusion of parentheses for method calls), but those are pretty minor and don't detract from the readability.

ThadHouse commented:
Is the plan that the RobotBase class managing these modes is single-threaded or multi-threaded? 

Definitely still single threaded, I don't expect that to change. Linear modes would still run in the main robot thread. It would need to be an external thread monitoring, that user code won't have access to (In fact, its possible its a completely separate process and not just a thread). And if the loop doesn't exit, the whole process would be torn down, not just the thread.

PeterJohnson commented:
> How do default commands work (or how would they be replaced)?

Default commands can still exist as they do today.  They will just only run when the command scheduler runs, which only happens in command-based modes.  We could change the example code to use a default command and it will still work as it does today.

> is having a separate robot base class for command-based that requires only command-based modes and calls the command scheduler itself across all modes

Yeah it’s mentioned in the trades section that potentially command-based should be a robot-wide thing, and you wouldn’t be able to mix command and non-command modes in a command-based robot.  That restriction might make command-based feel more cohesive, but would make it basically a completely separate project level thing you have to use, rather that being able to use it as a transitional thing in a project (which maybe isn’t realistic anyway due to subsystems etc, to your point).

jwbonner commented:
> We could change the example code to use a default command and it will still work as it does today.

In this case, does there need to be a way to create mode-specific default commands in addition to the current implementation? It seems like the example code was trying to demonstrate how a command like joystick drive could be bound to a specific mode. Global default commands are definitely still useful and should stick around, but it looked like the intention based on the example was to have a similar feature available in a mode-specific context.

> potentially command-based should be a robot-wide thing

This path makes sense to me overall. How can/should global periodic callbacks fit into this structure, given the utility of `robotPeriodic` for global telemetry? Command-based is built around a global periodic function already, and I guess the question is how much of that to expose to the user code (and in what way). That could take the form of a periodic function in the robot base class, an `addPeriodic` function in the robot base (which can't guarantee running in sync with the command scheduler), or something different (recommend that users make a command that ignores disable and is scheduled on init?).

The other question is whether something like this is in scope for the non-command variant of robot base. I think there's still an application for a periodic telemetry function regardless of mode, but that may just be impractical alongside e.g. linear modes.

KangarooKoala commented on design-docs/modes.md:
```diff
@@ -0,0 +1,622 @@
+Adding the following functions provides all the required functionality:
+- `int AddModeOption(string name, string category, string description, int flags)` – returns a unique ID identifying the mode (1:1 mapping to name).  Flags indicates which list the mode appears on.
+- `int RemoveModeOption(string name)` – returns the unique ID previously associated to that mode name
+ClearRoutines() – clears all modes
```
```suggestion
- `ClearRoutines()` – clears all modes
```

KangarooKoala commented on design-docs/modes.md:
```diff
@@ -0,0 +1,622 @@
+  private final Joystick joy = new Joystick(1);
+  private final Robot robot;
+
+  public AutoDriveStraight(Robot robot) {
```
```suggestion
  public Teleop(Robot robot) {
```

KangarooKoala commented on design-docs/modes.md:
```diff
@@ -0,0 +1,622 @@
+
+- FRC SendableChooser has a "default" option set by robot code.  Do we want something similar here or should it be 100% DS driven?  It's kind of nice to be able to set a default (e.g. via a `setDefaultAutonomousMode(String)` function in `Robot`), but also might be a little fragile since it's name based. If it's done via annotation, what happens if multiple annotations are marked as default?
+
+- Should it be possible to have multiple top-level Robot classes (e.g. for different robot configurations), and have that be selectable at the DS as well?  This is a bit ugly to support even if when the Robot object is passed into the mode constructor, because different Robot class types will only work with certain modes.
```
```suggestion
- Should it be possible to have multiple top-level Robot classes (e.g. for different robot configurations), and have that be selectable at the DS as well?  This is a bit ugly to support even if the Robot object is passed into the mode constructor, because different Robot class types will only work with certain modes.
```
I think that's how this should read?

KangarooKoala commented on design-docs/modes.md:
```diff
@@ -0,0 +1,622 @@
+
+- Should it be possible to have multiple top-level Robot classes (e.g. for different robot configurations), and have that be selectable at the DS as well?  This is a bit ugly to support even if when the Robot object is passed into the mode constructor, because different Robot class types will only work with certain modes.
+
+- Python--will decorators be able to work similarly to Java annotations for mode registration?
```
As [someone already commented in Discord](https://discord.com/channels/176186766946992128/368993897495527424/1350493490454331525), I doubt we could have a robust setup in Python for finding all classes that use a certain annotation. However, I'm not sure why we would want to use a metaclass-based registration instead of a decorator-based registration.

I unfortunately don't know the idiomatic way to handle the Python overloading, but something like this should work:
```py3
def Autonomous(cls_or_name: str | type, category: str = "default", description: str = ""):
  if isinstance(cls_or_name, str):
    def autonomous_mode_factory(cls: type):
      RobotBase.autonomous_mode_factory(lambda: cls(), cls_or_name, category, description)
      return cls
    return autonomous_mode_factory
  elif isinstance(cls_or_name, type):
    RobotBase.add_autonomous_mode(lambda: cls_or_name(), cls_or_name.__name__, category, description)
  else:
    # Handle invalid type here
    pass
```
(See [Python reference 8.8: Class definitions](https://docs.python.org/3/reference/compound_stmts.html#class-definitions) for details about the decorator process. Notably, no-arg at the decoration site means the decorator is called directly, while args at the decoration site means the decorator is called with the arguments and returns a callable that receives the class.)

Right now it uses `RobotBase.add_autonomous_mode()` which is currently non-static in the example. I'm not sure why it needs to be non-static, but if it does actually need to be non-static, then we could have a global dictionary or list that the modes get added to which the robot instance reads and registers when it is initialized.

KangarooKoala reviewed (commented):
At some point it'd also be nice to clean up group vs category, but it doesn't matter that much.

PeterJohnson commented on design-docs/modes.md:
```diff
@@ -0,0 +1,622 @@
+
+When a match starts, the FMS commands the driver station to enable the robot in autonomous mode for the autonomous period, followed by a short time disabled (typically 1-3 seconds), followed by enabling the robot in teleoperated mode for the teleoperated period, followed by disable.
+
+Separate from the driver station application, WPILib-provided as well as custom dashboard applications are commonly utilized by teams to provide a drop-down list for selection of autonomous routines.  Operators select the desired autonomous routine via the dashboard prior to the start of a match (or off-field testing of autonomous).  Some teams use other methods to select different autonomous rou ...
```
No noun required; what the text is saying is WPILib-provided dashboard applications as well as custom dashboard applications.

KangarooKoala replied:
Ah, that makes sense.

PeterJohnson commented on design-docs/modes.md:
```diff
@@ -0,0 +1,622 @@
+
+### Robot Code
+
+The standard WPILib team code structure derives a single team-written `Robot` class from an "periodic" robot base class.  This `Robot` class is constructed and run by the Java `main()` function.  After construction, the base class implementation runs a periodic loop (typically running on a 20 ms period, although that can be changed by the user) that reads the enabled state and teleop/auto/test mo ...
```
Changed "virtual" to "overrideable" everywhere.

PeterJohnson committed:
Various text cleanups

PeterJohnson committed:
Add start of migration guides with list of key differences

PeterJohnson committed:
Change all "category" to "group" for consistency

Starlight220 commented on design-docs/modes.md:
```diff
@@ -0,0 +1,638 @@
+
+Opmode classes are constructed when the user taps the Init button in the DS.  This works because hardware initialization processes never take more than a few seconds, and there's always a human lag between tapping Init and tapping Start.
+
+Hardware configuration is stored in an xml file that is consumed on APK startup, or when a different configuration is selected by the operator.  Consuming the xml file instantiates all objects (sensors, motors, servos, etc...) and stores them in a dictionary.  This dictionary then persists until either the robot is power cycled, or the operator selects a different configuration.  User code pulls  ...
```
```suggestion
Hardware configuration is stored in an XML file that is consumed on APK startup, or when a different configuration is selected by the operator.  Consuming the xml file instantiates all objects (sensors, motors, servos, etc...) and stores them in a dictionary.  This dictionary then persists until either the robot is power cycled, or the operator selects a different configuration.  User code pulls items out of the dictionary, typically in an initialization portion of an opmode, and then use throughout the opmode.  Notably, when the opmode finishes, or is stopped, the dictionary persists.
```

PeterJohnson committed:
Uppercase XML

Co-authored-by: Starlight220 <53231611+Starlight220@users.noreply.github.com>

alan412 commented on design-docs/modes.md:
```diff
@@ -0,0 +1,638 @@
+This document describes a standardized approach for operators to select different code to run for different robot modes of operation and for programmers to easily write code that creates these selection options.
+
+A note on terminology: the word "mode" is used in this document to describe this functionality.  While this word overlaps with the more general "modes" of robot operation (e.g. disabled, periodic, and auto), it's natural to describe the robot as having "several autonomous modes," and "in autonomous mode" (without referring to a specific one) refers to the general robot mode.
+
```
I think this confusion is probably a good enough reason for us to come up with another term than modes.

PeterJohnson replied:
"Routine" was about the only other option I came up with, unless we just want to stick with "OpMode" (which works, but might be confusing during the transition period, which is why I didn't use it).

alan412 commented on design-docs/modes.md:
```diff
@@ -0,0 +1,638 @@
+
+A clear enable/disable in the Driver Station that disables all robot actuators when disabled is a requirement because it is a safety-critical feature for FRC due to the size and power of FRC robots.
+
+The enabled "Init" step in the 2025 FTC SDK/DS is not a requirement due to anticipated rule changes.  However, performing initialization of modes while the robot is still disabled is a requirement for both FTC and FRC, as it's important for user code to be able to do expensive mode-specific operations (e.g. computing autonomous paths) prior to actually starting the match.
```
Does this just mean that for FTC, you don't have to be able to move actuators or does it mean more than this?

PeterJohnson replied:
Yes, the main difference compared to current FTC "init" is that you won't be able to move actuators in this state--the robot is still disabled.  You can still do other things (read sensors, do computations, etc), just not drive motors.

alan412 commented on design-docs/modes.md:
```diff
@@ -0,0 +1,638 @@
+
+- Modes may also be registered via annotation of functions (in the Robot class only) or via explicit function calls.  As C++ does not support annotations, function call registration is the only available method in that language.
+
+- For maximum flexibility, all code in the robot project has access to the enable/disable state, the overall robot teleop/auto/test mode, and the selected modes for teleop, auto, and test (even when the robot is disabled).
```
Why is this a good thing?   It feels like this could complicate things quite a bit.

PeterJohnson replied:
This is important for subsystems or other facilities (e.g. logging) that might want to change behavior based on the state of the robot.  We don't want to (and can't) hide this from teams, so we should provide a good API for it.

KangarooKoala replied:
It might also be worth clarifying that this is read-only access, so the complexity should be minimal. The library will already need to parse this data, so unless I'm missing something, this would only involve writing convenience methods that let teams read that data.

PeterJohnson replied:
And in general, the library itself will use these same methods (e.g. to implement the mode transition machinery).  So it's not any additional work over what we already have to implement.

alan412 commented on design-docs/modes.md:
```diff
@@ -0,0 +1,638 @@
+### RobotBase
+
+The `RobotBase` class is the base class for the user's `Robot` class.  It also implements the private library machinery for robot startup and robot execution (including creating and transitioning between modes in accordance with the mode lifecycle, as described in the following section).
+
```
This is not what I was thinking at all.   I was thinking of each "OpMode" (or whatever it gets named) having a class member that got a singleton instance of the "Robot" class.

PeterJohnson replied:
That's probably the biggest difference from current FTC opmodes.  The Robot instance is constructed first (by main()) and modes are constructed later.  Robot is effectively a singleton; it's a single object that persists through all modes that is provided to each mode's constructor (and the mode constructor stores it into a member variable).

alan412 reviewed (commented):
Overall, I am concerned about how complicated this is.   It does not feel approachable for a beginning programmer.

PeterJohnson committed:
Split command-based to separate document

Change command-based robots to use different Robot base class.
Remove CommandModes and integrate into this base class.
Add mode-sugared functions for Subsystem.setDefaultCommand().

PeterJohnson commented:
I split the command-based design into a separate design document.  Based on the initial feedback, I've changed it to be a separate robot base class (so it's not possible to intermix command-based and non-command-based modes).  I merged the CommandModes class into this new base class (so it's now less verbose to set up).  I also added support/sugar to set per-mode subsystem default commands.  I haven't added sugaring for joystick buttons yet, that's an open trade.

spacey-sooty commented on design-docs/modes.md:
```diff
@@ -0,0 +1,541 @@
+
+  // returns the current active mode;
+  // returns "" when the robot is disabled
+  public static String getActiveMode() {...}
```
Since disabledPeriodic from a selected mode runs while disabled should this return that mode?

PeterJohnson replied:
Interesting idea.  We'd have to change the DS behavior to do that, but it might be more consistent overall since we have the enabled bit still.

spacey-sooty replied:
If this is done I'd also change the other getActiveMode methods for consistency.

spacey-sooty commented on design-docs/commandbased-modes.md:
````diff
@@ -0,0 +1,184 @@
+The `Subsystem` interface adds an overload of `setDefaultCommand` to support creating per-mode default commands, and similarly a new `removeDefaultCommand` overload.  The non-`CommandMode` overloads are applied in all modes where a per-mode default has not been set.
+
+```java
+class Subsystem {
````
would we want to consider Resource for 2027?

spacey-sooty commented:
I don't see a way to set the default mode for a specific driverstation mode? Would that be something that could be added ie `RobotBase.setDefaultTeleoperatedMode` etc

ThadHouse commented:
At this point, I'm not sure we need a default. The way this is planned to be implemented at the DS level, its part of the direct control packets, so theres never a case one wouldn't be set. And the DS would save its last value.

spacey-sooty commented:
It would be nice to be able to enforce the default irrespective of what the driverstation had last selected.

PeterJohnson commented on design-docs/modes.md:
```diff
@@ -0,0 +1,541 @@
+
+# Unresolved Questions
+
+- FRC SendableChooser has a "default" option set by robot code.  Do we want something similar here or should it be 100% DS driven?  It's kind of nice to be able to set a default (e.g. via a `setDefaultAutonomousMode(String)` function in `Robot`), but also might be a little fragile since it's name based. If it's done via annotation, what happens if multiple annotations are marked as default?
```
The setting a default question is captured here as an unresolved one.  One issue is it's not obvious how it will work with annotations.  We'd also definitely have to work through all the same scenarios as dashboard sendable choosers re: when the default actually gets used... is it only when the DS is first started?  What about if the DS is an app or a physical device?

Maybe it might be better to have an annotation option to make some options bold/highlighted?  Or even allow setting a color code to make it more visually distinct?

We also might want different behavior for match vs non-match mode?

ThadHouse replied:
The main difference between sendable choosers and integrated into the DS is that the one integrated into the DS will _always_ be there. So does the default force change the DS? When does it do this? When the DS first starts? Every time code reboots? Every time the robot disconnects? Theres a lot more edge cases here. The only time the DS won't have a selection is the first time a project connects to the DS. Otherwise, the DS will always have something, and I feel like that being changed arbitrarily seems like it'd be more of a footgun then it should be.

Agreed that bolding/highlighting is a good idea.

SamCarlberg commented on design-docs/commandbased-modes.md:
```diff
@@ -0,0 +1,184 @@
+
+# Trades
+
+- Binding teleop joysticks is very verbose if different behavior is desired in different teleop modes.  Maybe add to CommandMode a separate event loop that's only active in that mode?  At the minimum, it may make sense to add CommandMode overloads to joysticks so users could write `driverController.x(teleop)` instead of `teleop.running.and(driverController.x())`?
```
From a code organization standpoint, declaring bindings for all modes in one place like the Robot constructor seems like it'd get very messy very quickly. Even the simple example above is showing some messiness. I think it's worth looking into better ways of declaring and separating mode-specific bindings; something that I like in Ruby (and which I used for epilogue configuration) is passing a lambda function as the last parameter that will accept the object to configure:

```java
CommandMode teleop = teleoperated("Teleop", mode -> {
  control.x().whileTrue(intake.intakeCommand());
  // etc...
});
```

This gives the framework control over when the configuration occurs, instead of it all happening at startup. For example, it means we could wipe all bindings when a mode exits and load all new bindings when a different mode starts by invoking the callback, without needing to worry about old bindings interfering with the new mode. Though that would effectively act as if everything in the lambda had a `mode.running.and(...)` wrapping everything, so more thought is definitely needed.

PeterJohnson replied:
Currently, command-based does all bindings up-front, and deferred binding seems like a pretty fundamental change.  The risk of deferring it is that binding errors couldn’t be caught until later, and throwing exceptions at mode start is a good way to have feels-bad moments in a match.

We can certainly provide better organization than shown here by breaking it into separate functions—for example we could move the mode bindings into the Autos class and create a Teleops class with functions that do the teleop bindings?

PeterJohnson replied:
I split up the mode creation functions into multiple functions in Robot, and put the factory functions back into CommandModes.

jeversmann replied:
> The risk of deferring it is that binding errors couldn’t be caught until later, and throwing exceptions at mode start is a good way to have feels-bad moments in a match.

If the robot is on the field with binding errors, whether they occur at the start of teleop or when the robot first turns on, then it's already too late to deploy working code.

PeterJohnson committed:
Rename commandbased-modes to modes-commandbased

PeterJohnson committed:
Move factory functions back to CommandModes

PeterJohnson committed:
Commandbased: Refactor mode creation into functions

PeterJohnson committed:
Improve/simplify non-command example code

KangarooKoala commented on design-docs/modes-commandbased.md:
```diff
@@ -0,0 +1,205 @@
+
+# Trades
+
+- Binding teleop joysticks is very verbose if different behavior is desired in different teleop modes.  Maybe add to CommandMode a separate event loop that's only active in that mode?  At the minimum, it may make sense to add CommandMode overloads to joysticks so users could write `driverController.x(teleop)` instead of `teleop.running.and(driverController.x())`?
```
Even though we have the jinja setup to make it no so much work on our end, I'm still worried about the increase in overloads from adding another optional parameter. More overloads makes it harder to see what's available in the class and can also make it more intimidating (especially for newer programmers with less support).

Instead, we can add a method to `Trigger` that will constrain the `Trigger` to only run in the active mode.
```java
public Trigger whileRunning(CommandMode mode) {
  // Short-circuit when we aren't running the provided mode
  return mode.running.and(this);
}
```

Then we can do `driverController.x().whileRunning(teleop)`, which is a little longer than `driverController.x(teleop)` but also reads a lot more naturally. (It's still less verbose than `teleop.running.and(driverController.x())`.)

PeterJohnson committed:
Rename modes to opmodes

ArchdukeTim commented on design-docs/opmodes.md:
```diff
@@ -0,0 +1,516 @@
+- `int RemoveOpModeOption(string name)` – returns the unique ID previously associated to that opmode name
+- `void ClearOpModeOptions()` – clears all opmodes
+- `int GetActiveOpMode()` - gets the currently active opmode ID; returns 0 when the robot is disabled
+- `int GetSelectedAutonomousOpMode()` - gets the opmode ID currently selected for autonomous; valid when the robot is disabled; returns 0 if unknown or no selection
```
What does this return if you have  separate auto op-modes selected for just auto vs the auto portion of match?

PeterJohnson replied:
The protocol only has one location to hold what this function returns.  So this is really more of DS behavior question when you change from auto to match.

ArchdukeTim commented on design-docs/opmodes.md:
```diff
@@ -0,0 +1,516 @@
+- There is no `@Disabled` annotation for opmodes; the opmode annotation can simply be commented out to achieve the same result.
+
+# Drawbacks
+
```
lol

ArchdukeTim commented on design-docs/opmodes.md:
```diff
@@ -0,0 +1,516 @@
+
+# Trades
+
+- Overall naming: mode? routine? opmode?
```
Mode is definitely a standard nomenclature (e.g "What auto modes does your team have?"). 'Teleop OpMode' also doesn't roll off the tongue, and will likely be shortened to 'teleop mode' in the vernacular.

PeterJohnson replied:
We have literally gone back and forth several times on what to call this.  Just calling it “mode” has potential confusion as well.

ArchdukeTim commented on design-docs/opmodes.md:
```diff
@@ -0,0 +1,516 @@
+
+- Overall naming: mode? routine? opmode?
+
+- Naming of opmode functions? start-periodic-end, vs init-periodic-exit (2025 FRC PeriodicRobot), vs init-execute-end (2025 FRC Command), vs init-start-loop-stop (2025 FTC OpMode; note init behaves like the constructor here)
```
suggestion: `onEnable`, `onDisable`. `init` will likely be confused with constructors, especially teams transitioning from `robotInit`, and we should try to avoid setting teams up for hardware initialization in this method accidentally.

ArchdukeTim replied:
Or maybe it won't be a huge problem since hardware initialization should (needs?) happen outside of the opmode in the robot class.

ArchdukeTim commented on design-docs/opmodes.md:
```diff
@@ -0,0 +1,516 @@
+
+- Naming of opmode functions? start-periodic-end, vs init-periodic-exit (2025 FRC PeriodicRobot), vs init-execute-end (2025 FRC Command), vs init-start-loop-stop (2025 FTC OpMode; note init behaves like the constructor here)
+
+- For matches, should we construct teleop at the same time as auto?  If we do that, we probably need a disabledStart() or 2025 FTC opmode style init() that's run on teleop after auto completes, and don't run disabledPeriodic for both simultaneously.
```
Will there be support for changing the op-mode at the end of auto based on conditions? Or will it always be DS selection based?

PeterJohnson replied:
Always DS driven.  We could/should support changing the DS selection during a match (eg just in case the user picked the wrong one).  Keep in mind this is all just software though; if the user wants to ignore the opmode setting from the DS entirely they can, we just won’t have a standard way to do that.

ArchdukeTim commented on design-docs/opmodes.md:
```diff
@@ -0,0 +1,516 @@
+
+# Unresolved Questions
+
+- FRC SendableChooser has a "default" option set by robot code.  Do we want something similar here or should it be 100% DS driven?  It's kind of nice to be able to set a default (e.g. via a `setDefaultAutonomousOpMode(String)` function in `Robot`), but also might be a little fragile since it's name based. If it's done via annotation, what happens if multiple annotations are marked as default?
```
Annotation processing could detect at compile-time, right?

PeterJohnson replied:
We could do that, yes.  I just would want to avoid doing any compile time code generation; we’ve found that to not play nicely with Gradle/IDE support.

KangarooKoala replied:
Would a compile time code checker be feasible though? I'd guess that we'd run into less issues since we wouldn't need to consume outputs of the annotation processing.

PeterJohnson commented on design-docs/opmodes-commandbased.md:
````diff
@@ -0,0 +1,205 @@
+```java
+class Subsystem {
+  // Sets a default command for this subsystem active only in the given opmode.
+  default void setDefaultCommand(CommandOpMode opmode, Command defaultCommand);
````
I did it because it looked nicer to my eyes as well.  Good style question though!

nihonjinrxs commented on design-docs/opmodes-commandbased.md:
```diff
@@ -0,0 +1,205 @@
+
+# Motivation
+
+See [opmodes](opmodes.md) for the motivation for adding operator selectable opmodes to the core robot structure.  For command-based programs, the entire robot program is structured around command-based subsystems and commands tied to robot states and inputs, with everything constructed during initialization, so for more natural integration with the rest of the command-based framework, a different ...
```
Can you explain more about the problem caused by attempting to use an annotation-based approach in a command-based project? I'm not sure I understand why this wouldn't work and feel fairly natural. Is there some compile-time vs runtime difference I'm missing here?

PeterJohnson replied:
Using annotations in what way?  It wouldn't make sense to use them on classes like for the non-command-based modes, because all command-based things are set up during robot construction.  Maybe function annotations?  The reason we don't allow a mix of non-command and command-based opmodes is because of the assumptions surrounding subsystem periodic and the scheduler always being called (see @jwbonner's earlier comment).

nihonjinrxs commented on design-docs/opmodes-commandbased.md:
```diff
@@ -0,0 +1,205 @@
+# Trades
+
+- Binding teleop joysticks is very verbose if different behavior is desired in different teleop opmodes.  Maybe add to CommandOpMode a separate event loop that's only active in that opmode?  At the minimum, it may make sense to add CommandOpMode overloads to joysticks so users could write `driverController.x(teleop)` instead of `teleop.running.and(driverController.x())`?
+- Is the `beforeScheduler()` / `afterScheduler()` split beneficial?  Or should we just have `robotPeriodic()` showing the `CommandScheduler.run()` code in it like we do in 2025?
```
From my perspective, the `beforeScheduler()` / `afterScheduler()` split just creates extra functions. It's fairly clear what they are doing, but I don't think it's bad to keep the current approach with `CommandScheduler.run()` in `robotPeriodic()` is it is in 2025.

PeterJohnson replied:
I think if we do that, we need to clearly define whether the subsystem periodic() is called before or after the scheduler.  Right now it's not well defined and is a point of confusion for teams.

nihonjinrxs commented on design-docs/opmodes-commandbased.md:
```diff
@@ -0,0 +1,205 @@
+
+- Binding teleop joysticks is very verbose if different behavior is desired in different teleop opmodes.  Maybe add to CommandOpMode a separate event loop that's only active in that opmode?  At the minimum, it may make sense to add CommandOpMode overloads to joysticks so users could write `driverController.x(teleop)` instead of `teleop.running.and(driverController.x())`?
+- Is the `beforeScheduler()` / `afterScheduler()` split beneficial?  Or should we just have `robotPeriodic()` showing the `CommandScheduler.run()` code in it like we do in 2025?
+- Related: should we remove the built-in `periodic()` from the `Subsystem` interface?  Split it into before/after?
```
I'd like to strongly disagree with the idea of removing the `periodic()` method. There's an approach to managing robot subsystem code that my team uses, and that I've heard from several other teams they use in varying degrees, where subsystems operate as state machines via their `periodic()` functions, controlling actuators according to the subsystem state, and commands set states on those subsystems. Entirely removing the `periodic()` method from the `Subsystem` interface would make that approach impossible, and that approach has made for very readable and rather performant code over the last couple seasons, even on RoboRIO 1.

I think I'd be fine with splitting the `periodic()` method into a `beforeScheduler()` / `afterScheduler()` split, which would allow for doing things like getting updated sensor values from hardware prior to the command scheduler run, then making decisions about what to do with actuators and how to update the subsystems state, as well as doing things like telemetry publishing, after the scheduler run. This would require some documentation, I think, but is a reasonable transition to make.

PeterJohnson replied:
Except the periodic() function can be called explicitly from robotPeriodic() if desired.  The fact it's in the interface just makes it more automagic because the scheduler.run() method calls it for you.

thenetworkgrinch commented on design-docs/opmodes-commandbased.md:
```diff
@@ -0,0 +1,205 @@
+
+    // Control the drive with split-stick arcade controls
+    m_drive.setDefaultCommand(
+        opmode,
```
This is extremely verbose and often pointless to specify since operator controls are in teleop opmodes. Can’t there be a way to add commands per-opmode if defined like
`opmode.add(controller.x().whileTrue(()->drive.drive(…)),…)` and default to all of the teleop/test opmodes. This way seems less verbose and more structured.

PeterJohnson replied:
How would the implementation of that work?  It might be possible by some way injecting an additional condition into the trigger creation/binding behind the scenes, but that feels somewhat fragile.

thenetworkgrinch replied:
It could work if the Triggers were bound to an opmode inside of the class which can be checked in the command scheduler periodic?

The opmode can call a `Trigger.setOpMode(OpMode)` inside the `OpMode.add` function. Triggers can be constructed with an optional opmode filter which stores the opmode(s) it runs under.

jeversmann commented:
I would love to see a top level design that allows for a smooth progression from linear/periodic opmodes into command based opmodes without requiring those concepts to exist in separate projects. Right now the FRC students I've worked with are command based all the time because we force them to be from the beginning, but FTC students have a progression from "all of our code is in one file" to a codebase with 4+ autos and at least one teleop that are sharing code but without the benefits of a command based architecture. I would love to be able to walk students through upgrading a codebase in place from linear/periodic opmodes to having everything be command based instead of having to transition code across over to a new project or break existing opmodes.

I think the requirement of a `Robot` singleton is a great starting point since hardware configuration code is already conceptually different from opmode code and this immediately gets teams over the hurdle of sharing hardware configuration between teleop and auto.
From there I would love to be able to introduce `InstantCommand` as a way to share robot functionality between teleop and auto, especially for markers on paths¹. At this stage, students could construct and invoke commands in their teleop code using `if/else` button bindings inside of linear/periodic opmodes rather than interacting with the scheduler, but they could be writing `Command` classes that could continue to work as their code increases in complexity.
Concepts like `Subsystem` requirements and the full `Command` lifecycle could come up incrementally while working on more complex autos, and students could make their own decision about when they're ready to hand teleop over to the command scheduler.

I understand that FRC teams are used to having `CommandScheduler` run at all times after robot initialization, but in simpler codebases I would love to have an abstract `CommandOpMode` that lives alongside linear/periodic opmodes and allows for bindings to be set up which live only during a specific robot mode and hides the complexity of the scheduler.
```java
@Teleoperated
public class Teleop extends CommandOpMode {
  public Teleop(Robot robot) {
    var driverController = new CommandXboxController(1);

    // Control the drive with split-stick arcade controls
    robot.drive.setDefaultCommand(
        robot.drive.arcadeDriveCommand(
            () -> -driverController.getLeftY(), () -> -driverController.getRightX()));

    // Deploy the intake with the X button
    driverController.x()).onTrue(robot.intake.intakeCommand());
    // Retract the intake with the Y button
    driverController.y()).onTrue(robot.intake.retractCommand());
  }
}
```
This would simplify some of the proposed binding code, with the drawback that it adds additional complexity to managing the state of the robot while it isn't enabled. It seems to me that where teams need bindings that persist between teleop/auto/disabled modes that justifies shifting that complexity up into the `Robot` definition and being more verbose if necessary. If there's a need for a separate `CommandRobot` that's incompatible with linear/periodic opmodes in order to match the existing command based capabilities of FRC software I think that's fine, but I would want it to be an incremental increase in complexity available to a codebase which already has `Subsystem` and `Command` classes defined and working with the simpler `Robot` interface.

 ¹ I'm really excited for controller hosted path planning being available to teams by default. Upgrading an FTC codebase to Road Runner is a barrier that prevents many teams in my area from having competitive autos, and I would love to have similar path building & following tools available to students without having to mess with external dependencies. However, a lot of the accessibility that the PathPlanner provides to FRC relies on `NamedCommands` and I would hate for teams to lock themselves off from making autos with a GUI based tool because their teleop opmode is linear/periodic.

PeterJohnson commented:
Pedagogically, linear and periodic have a different robot-wide structure and expectations than command-based.  Command-based is designed to have nearly everything set up at robot construction (connecting commands to triggers and defining command subsystem dependencies), and also has a very specific definition of what a "subsystem" needs to be in order for command dependencies to logically work.  There's also recommended design patterns unique to command-based (such as using factory functions to create commands in subsystems) that would likely be confusing with periodic/linear modes mixed in.  I would be concerned that a team starting from a periodic design approach and trying to gradually transition to command based within the same robot project would really struggle to understand why their subsystems they used when developing their periodic design don't map to a command-based approach, or things will break in surprising ways in linear or periodic code because the scheduler isn't calling subsystem periodic behind the scenes in those cases, and in general they will end up with more difficult to understand and less maintainable code with poor design patterns by going down this road.

Implementation wise, having a CommandOpMode along the lines of what you propose to allow mixing of the two is where I started, but changed it due to Jonah's comment (https://github.com/wpilibsuite/allwpilib/pull/7863#issuecomment-2726749673).  I think the disabled case is probably the most challenging to think about re: running the scheduler.  We could call it when a command-based mode is *selected* (by running it from CommandOpMode.disabledPeriodic), but if a team wants it *always* to be called (even before a command is selected), there's not a good way to do that--they can't put it in robotPeriodic or disabledPeriodic, because then it would run twice.  I suppose we could add a robot-wide noOpModeSelectedPeriodic type of call to handle that specific case, but that feels like it adds even more complexity to an already complex system.

One other point is that the approach where command-based constructs everything up front benefits teams because various types of construction errors can be handled by throwing exceptions, and that's okay, because it's very obvious that your code didn't start.  Deferring that work to the opmode selection means we will want to avoid that (because we don't want to crash code e.g. when the auto mode is selected or at the beginning of teleop), which will make errors more difficult for teams to discover and diagnose.

qwertychouskie commented:
Having a single `Robot` is likely to be very problematic for FTC.

In FTC, it's common to have multiple robots with reasonably substantial differences through the course of the season.  For example, might have your `RobotIn3DaysRobot`, your `LeagueMeetRobot`, and your `ShinyNewRebuildRobot`.  In the current control system, you would have e.g. `3dayTeleop` and `3dayAutonomous` OpModes with `Robot robot = new RobotIn3DaysRobot()` in them, with `RobotIn3DaysRobot` providing all the `Motor` and `Servo` and `Sensor` and such objects; and so on for your `LeagueMeetRobot` and your `ShinyNewRebuildRobot`.

With only a single `Robot` allowed, would we have to have an entirely separate codebase for each robot?  We can't just delete the old `Robot` class and associated OpModes the moment we start on the new one, as even though e.g. `LeagueMeetRobot` is unlikely to significantly change while working on `ShinyNewRebuildRobot`, the `LeagueMeetRobot` will almost certainly still see minor changes for the sake of the drivers, until `ShinyNewRebuildRobot` is complete enough for the drivers to start driving that robot instead.

And that's not even to count the various test `Robot`s used with test OpModes for testing various things, e.g. throwing a new odometry sensor on a spare strafer chassis to test it and make sure it's working as expected.

Honestly, having `Robot` on top of the hierarchy instead of the OpModes just feels like the wrong order.

TheComputer314 commented:
```java
@RobotsManager
public class RobotsManager {
    private final Robot oldRobot;
    private final Robot newRobot;
    public RobotsManager() {
        this.oldRobot = new OldRobot();
        this.newRobot = new NewRobot();
    }
}
```
Wouldn't this defeat the whole purpose of different hardware setups? When `RobotsManager` gets initialized, ***both*** `Robot` classes get instantiated here, which can mess things up. (Most likely via a crash)

If a team needs different hardware setups (IE, slab bot testbed) then we can add the ability to select which `Robot` class to load on code initialization and have the user select in code via some constants or identifying feature. (See Peter's comment) (`public static boolean TEST_BOT = true;`, `RobotController.getSerialNumber().equals("123ABC");`, etc) However, unloading a `Robot` then loading in a new one in runtime is a ***complete non-starter*** due to long initialization time and memory/physical safety challenges of making sure everything in the old `Robot` got unloaded and cleaned up properly. This can also already be done in current FRC via the [IO layer system,](https://docs.advantagekit.org/data-flow/recording-inputs/io-interfaces) just that IO layers do it at a lower, more granular level than replacing the whole `Robot` class.

If the robot hardware architecture is different enough that code can't just be commented out or swapped via IO layers/a different `Robot` class, then IMO, it would be less hassle to maintain a separate branch/robot project.

I also don't see the use for having multiple, highly different robot architectures available simultaneously. If build team's working on a rebuild, it's safe to assume the old bot's not gonna get touched, and it's gonna be relegated to driver practice. If any code changes arise during those driver practice in this scenario of multiple bots, and that code change needs to be carried over to the new bot, copy paste or cherry pick isn't hard.

qwertychouskie commented:
> ```java
> @RobotsManager
> public class RobotsManager {
>     private final Robot oldRobot;
>     private final Robot newRobot;
>     public RobotsManager() {
>         this.oldRobot = new OldRobot();
>         this.newRobot = new NewRobot();
>     }
> }
> ```
> 
> Wouldn't this defeat the whole purpose of different hardware setups? When `RobotsManager` gets initialized, _**both**_ `Robot` classes get instantiated here, which can mess things up. (Most likely via a crash)
> 
> If a team needs different hardware setups (IE, slab bot testbed) then we can add the ability to select which `Robot` class to load on code initialization and have the user select in code via some constants or identifying feature. (See Peter's comment) (`public static boolean TEST_BOT = true;`, `RobotController.getSerialNumber().equals("123ABC");`, etc) However, unloading a `Robot` then loading in a new one in runtime is a _**complete non-starter**_ due to long initialization time and memory/physical safety challenges of making sure everything in the old `Robot` got unloaded and cleaned up properly. This can also already be done in current FRC via the [IO layer system,](https://docs.advantagekit.org/data-flow/recording-inputs/io-interfaces) just that IO layers do it at a lower, more granular level than replacing the whole `Robot` class.
> 
> If the robot hardware architecture is different enough that code can't just be commented out or swapped via IO layers/a different `Robot` class, then IMO, it would be less hassle to maintain a separate branch/robot project.
> 
> I also don't see the use for having multiple, highly different robot architectures available simultaneously. If build team's working on a rebuild, it's safe to assume the old bot's not gonna get touched, and it's gonna be relegated to driver practice. If any code changes arise during those driver practice in this scenario of multiple bots, and that code change needs to be carried over to the new bot, copy paste or cherry pick isn't hard.
> 
> > 1. User accidentally selects OpModeA that uses e.g. RobotA
> > 2. Robot gives error or does wrong thing or whatever
> > 3. User selects OpModeB that uses RobotB
> 
> This can be entirely avoided if the OpModes that don't use `RobotB` don't get shown at all, ex by never instantiating `RobotA`.

Yeah the `RobotsManager` idea has some kinks.  Just changing the `Robot` class being created in `main` like guinawheek posted works fine.

```
/**
 * Do NOT add any static variables to this class, or any initialization at all. Unless you know what
 * you are doing, do not modify this file except to change the parameter class to the startRobot
 * call.
 */
public final class Main {
  private Main() {}

  /**
   * Main initialization function. Do not perform any initialization here.
   *
   * <p>If you change your main robot class, change the parameter type.
   */
  public static void main(String... args) {
    RobotBase.startRobot(Robot::new); // <-- change Robot::new with DifferentRobot::new at compile time
  }
}
```
This should probably be well documented though, as it's behavior that many FTC teams will be looking for.

Froze-N-Milk commented:
I generally find the idea of a robot class an antipattern in combination with an opmode, when an opmode should describe the linking and binding of subsystems and commands. At what point is the opmode class fairly pointless, serving only to provide a name and a categorisation? (seems like command opmodes as currently modelled don't actually exist as classes, and solely do this). Generally in ftc the model is: opmode -> creates / pulls in subsystems -> links commands to gamepads / does other human input linking with robot action. i have a hard time not seeing robot class and opmode as interchangeable here, just with a different perspective.

BeepBot99 commented:
An idea:

What if OpModes weren't created in the `Robot` class and were instead in the entry-point (similar to how ASP.NET works, but of course it doesn't have OpModes)? But the `Robot` would still contain all the hardware.

So `main` would look something like this:

```java
RobotBuilder builder = new RobotBuilder();
builder.addRobot(new MyRobot());
builder.addAutonomous("Simple Auto").running.whileTrue(Autos.simpleAuto(builder.robot));
CommandOpMode opMode = builder.addTeleoperated("TeleOp");
// and all the rest
// then:
builder.build().run();
```

Advantages to this approach:

- It is much more clear exactly what `addAutonomous` and `addTeleoperated` are doing. With having them be static methods in `CommandOpModes`, it's not very clear that they have a side effect (registering the OpMode).

Disadvantages to this approach:

- If it's all in main then having multiple robots would suck, although you could always just make the entire thing a static method in a different class and call it from main.

Also, a possible idea for the verboseness of gamepads in specific OpModes:

```java
opMode.bindWhileRunning(
  driverControlled.x(), onTrue(intake.intakeCommand()),
  driverControlled.y(), onTrue(intake.retractCommand())
);
```

Additionally, possibly a way to do command OpModes as classes by extending a `CommandOpModeHelper` base class and then registering using `CommandOpModes.addAutonomous(MyCommandOpMode())` (or `builder.addAutonomous(...)`.

In `CommandOpModeHelper`, you would pass the name of the OpMode, probably to the super constructor. There would also be an easy way to write start and periodic code in the OpMode, that is turned into a command. Also, there could be methods that help with bindings for that OpMode to reduce the verboseness, something like:

```java
// in the constructor:
trigger(driverControlled.x()).onTrue(...);
```

KangarooKoala commented:
After over an hour of discussion in Discord (starting [here](<https://discord.com/channels/176186766946992128/1365422855378178048/1366930589835460638>) and ending somewhere around [here](<https://discord.com/channels/176186766946992128/1365422855378178048/1366948174660501556>)), the group reached the following consensus about how to cleanly restrict triggers to particular opmodes (with several open questions remaining).

```md
Commands contexts:

* A `Context` represents a condition for a trigger being active. Either via inheritance, overloads, a public final field, sugar is provided to use a `CommandOpMode` as a `Context` which restricts bindings to when the opmode is currently running. (`CommandOpMode.running`)

      public class Context {
        public static Context always();
        public static Context disabled();
        public static Context enabled();
        public static Context auto();
        public static Context teleop();
        public static Context test();

        public Context(BooleanSupplier);

        public Context and(BooleanSupplier);
        public Context or(BooleanSupplier);

        public boolean isActive();
      }

  * Open question: Should `Context` extend `BooleanSupplier`? Concern: `Condition.and(Context)` versus `Condition.in(Context)` may be confusing. Would we want `Condition.and(Context)` to behave identically to `Condition.in(Context)`? (This would include returning a `Trigger` instead of a `Condition`.) Is `Condition.or(Context)` something we want to allow?
  * Open question: If yes to the above, should context objects be of type `BooleanSupplier` or `Condition` instead of having a separate `Context` type? Note that the purpose of mandatory `Context`s is to avoid accidentally forgetting them, which would be muddied if the type is `BooleanSupplier` or `Condition` instead of `Context`.
  * Open question: Should `Context.always()` be renamed to `Context.all()`? The original discussion used `Context.all()` (and sometimes `OpMode.all()`), but never explicitly discussed the name.

* A `Condition` represents a boolean condition, but cannot make any bindings. Its primary purpose is to provide convenience boolean logic methods, with clean interactions with `Trigger`s.

      public class Condition extends BooleanSupplier {
        public Condition(BooleanSupplier);

        @Override
        public boolean getAsBoolean();

        public Trigger in(Context);

        public Condition and(BooleanSupplier);
        public Trigger and(Trigger);
        public Condition or(BooleanSupplier);
        public Trigger or(Trigger);
        public Condition negate();
        public Condition debounce(double);
        public Condition debounce(double, Debouncer.DebounceType);
      }

  * Open question: Should we encourage using arbitrary `Condition`s as `Context`s by adding `public Context asContext()`?

* A `Trigger` represents a condition (combination of `Context` and base condition) that can make command bindings. It is associated with a `Context`. A `Trigger` with an inactive `Context` does not evaluate its base condition and returns false. (This applies to both using the `Trigger` as a `BooleanSupplier` and using the `Trigger` to make command bindings.)

      public class Trigger implements BooleanSupplier {
        public Trigger(Context, BooleanSupplier);
        public Trigger(Context, EventLoop, BooleanSupplier);

        @Override
        public boolean getAsBoolean();

        public Trigger and(BooleanSupplier);
        public Trigger or(BooleanSupplier);
        public Trigger negate();
        public Trigger debounce(double);
        public Trigger debounce(double, Debouncer.DebounceType);

        public Trigger onChange(Command);
        public Trigger onTrue(Command);
        public Trigger onFalse(Command);
        public Trigger whileTrue(Command);
        public Trigger whileFalse(Command);
        public Trigger toggleOnTrue(Command);
        public Trigger toggleOnFalse(Command);
      }

  * Open question: Should `Trigger` extend `Condition`, or should it just have parallel APIs? Anything using `Condition` should probably ask for `BooleanSupplier` instead anyways for generality. On the other hand, conceptually a `Trigger` is a `Condition`. However, `Trigger.in(Context)` doesn’t make that much sense. Furthermore, we need to override all of the methods to have the right return type anyways.
  * Open question: How should we resolve combining `Trigger`s with different contexts (or event loops)?
  * Open question: Should we encourage using arbitrary `Trigger`s as `Context`s by adding `public Context asContext()`?
  * Open question: Should we make any distinction between the Context, the base condition, and the combined condition? This depends on how we resolve combining Triggers with different contexts.

* The reason for the `Condition`/`Trigger` split is to require a `Context` for making command bindings, preventing bindings from activating at unexpected times.

* The command controller classes (e.g., `CommandXboxController`) have a mandatory context parameter for their constructors. Otherwise, they have the same overload set as before.

      public class CommandGenericHID {
        // Removed public CommandGenericHID(int);
        public CommandGenericHID(int, Context);

        // Otherwise, the API is the same, with the Context
        // being provided to the Triggers.
      }

* `CommandOpMode` continues to provide public final `Trigger` fields. Binding commands (e.g., auto path preloading) to stages of op mode execution is still important.
  * Open question: What should be the Context of the provided Triggers? Some starting options are: Context.always(), new Context(opmode.selected), and the trigger’s condition. If the Context of a Trigger is never exposed separately from the overall condition, then this question is moot, but that is not a given.

* There are two recommended structures for bindings that are common to multiple opmodes:
  * Define a method with an opmode parameter that uses that parameter for all contexts, and invoke/call that method for each opmode with the shared bindings.
  * Separate from opmode-specific bindings, make the bindings with the context being a combination of all opmodes with the shared bindings.
```

KangarooKoala committed:
Add command contexts design doc (#18)

PeterJohnson commented:
Closing; an updated version of the non-command based design doc was merged with the implementation.  The command based one will similarly come in with its implementation.
